test-interface/src/main/scala/org/scalajs/testinterface/HTMLRunner.scala

```diff
@@ -20,130 +20,239 @@ import sbt.testing._
+
+    class Handler extends EventHandler {
+      private[this] var _hasErrors = false
+
+      def handle(event: Event): Unit = {
+        val status = event.status
+        _hasErrors ||= isErrorStatus(status)
+        counts(status) += 1
+      }
+
+      def hasErrors: Boolean = _hasErrors
+    }
   }
 
   def main(): Unit = {
-    val runResults = for {
-      (framework, taskDefs) <- TestDetector.detectTests()
-    } yield {
-      logger.debug(s"Detected ${taskDefs.size} tests for ${framework.name}")
-      val result = runTests(framework, taskDefs)
+    val allTests = TestDetector.detectTests()
 
-      // Report results.
-      result.filter(_.nonEmpty).foreach(r =>
-          logger.info(s"${framework.name()} reported $r"))
+    val ui = new UI(testCount = allTests.map(_._2.size).sum)
 
-      result.map(_ => ())
+    val oks = for {
+      (framework, taskDefs) <- allTests
+    } yield {
+      runTests(framework, taskDefs, ui)
     }
 
     // Report event counts.
-    Future.sequence(runResults).foreach { (_: Seq[Unit]) =>
-      import EventCounter.counts
+    Future.sequence(oks).map(and).foreach(ui.done)
+  }
+
+  private def runTests(framework: Framework,
+      taskDefs: Seq[TaskDef], ui: UI): Future[Boolean] = {
+    def runAllTasks(tasks: Seq[Task]): Future[Boolean] = {
+      val oks = tasks.map { task =>
+        for {
+          (ok, newTasks) <- scheduleTask(task, ui)
+          newOk <- runAllTasks(newTasks)
+        } yield ok && newOk
+      }
+
+      Future.sequence(oks).map(and)
+    }
+
+    val runner = framework.runner(Array(), Array(), classLoader)
+    val tasks = runner.tasks(taskDefs.toArray)
+
+    for (ok <- runAllTasks(tasks)) yield {
+      val resultStr = runner.done()
+      if (resultStr.nonEmpty)
+        ui.reportFrameworkResult(ok, framework.name, resultStr)
+      ok
+    }
+  }
+
+  private def scheduleTask(task: Task, ui: UI): Future[(Boolean, Array[Task])] = {
+    val promise = Promise[(Boolean, Array[Task])]
+
+    val uiBox = new ui.TestTask(task.taskDef.fullyQualifiedName)
+    val handler = new EventCounter.Handler
+
+    // Don't use a Future so we yield to the UI event thread.
+    js.timers.setTimeout(0) {
+      task.execute(handler, Array(uiBox.logger), { newTasks =>
+        val ok = !handler.hasErrors
+
+        uiBox.done(ok)
+        promise.success((ok, newTasks))
+      })
+    }
+
+    promise.future
+  }
 
-      // Insert <hr>
-      UI.hr()
+  private class UI(testCount: Int) {
+    // UI Elements.
+
+    // Top level container. Mainly to prevent code under test from accidentally
+    // modifying our UI.
+    private val container = dom.document.body.newElement()
+    private val rootBox = new RootBox(testCount)
+
+    // State.
+    private var _done = false
+
+    updateCounts()
+
+    def done(ok: Boolean): Unit = {
+      _done = true
+      rootBox.box.done(ok)
+      rootBox.counterLine.className = "log " + statusClass(ok)
```

Comment:
Looks like these two lines should be in a method `RootBox.done(ok)` to improve encapsulation.